third_party/move/tools/move-asm/src/syntax.rs

```diff
@@ -0,0 +1,783 @@
+                };
+                Ok((name, abs))
+            },
+            ",",
+        )?;
+        self.expect_special(">")?;
+        Ok(result)
+    }
+
+    fn visibility(&mut self) -> AsmResult<Visibility> {
+        Ok(if self.is_soft_kw("public") {
+            self.advance()?;
+            Visibility::Public
+        } else if self.is_soft_kw("friend") {
+            self.advance()?;
+            Visibility::Friend
+        } else {
+            Visibility::Private
+        })
+    }
+
+    fn local_decl(&mut self) -> AsmResult<Local> {
+        let loc = self.next_loc;
+        let name = self.ident()?;
+        self.expect_special(":")?;
+        let ty = self.type_()?;
+        Ok(Local { loc, name, ty })
+    }
+
+    fn argument(&mut self) -> AsmResult<Argument> {
+        if self.is_special("<") {
+            self.advance()?;
+            let ty = self.type_()?;
+            self.expect_special(">")?;
+            Ok(Argument::Type(ty))
+        } else if self.is_value() {
+            let val = self.value()?;
+            Ok(Argument::Constant(val))
+        } else {
+            let pid = self.partial_ident()?;
+            let targs = self.type_args_opt()?;
+            Ok(Argument::Id(pid, targs))
+        }
+    }
+
+    fn is_argument(&self) -> bool {
+        self.is_value() || self.is_partial_ident() || self.is_special("<")
+    }
+
+    fn instr(&mut self) -> AsmResult<Instruction> {
+        let mut loc = self.next_loc;
+        let mut name = self.ident()?;
+        let label = if self.is_special(":") {
+            self.advance()?;
+            let label = Some(name);
+            loc = self.next_loc;
+            name = self.ident()?;
+            label
+        } else {
+            None
+        };
+        let args = if self.is_argument() {
+            self.list(Self::argument, ",")?
+        } else {
+            vec![]
+        };
+        // Extend loc to include args
+        let loc = Loc::new(loc.start(), self.previous_loc.end());
+        Ok(Instruction {
+            loc,
+            label,
+            name,
+            args,
+        })
+    }
+
+    fn is_fun(&self) -> bool {
+        self.is_soft_kw("fun") || self.is_soft_kw("public") || self.is_soft_kw("friend")
+    }
+
+    fn fun(&mut self) -> AsmResult<Fun> {
+        let visibility = self.visibility()?;
+        self.expect_soft_kw("fun")?;
+        let loc = self.next_loc;
+        let name = self.ident()?;
+        let type_params = self.type_params()?;
+        self.expect_special("(")?;
+        let params = if self.is_ident() {
+            self.list(Self::local_decl, ",")?
+        } else {
+            vec![]
+        };
+        self.expect_special(")")?;
+        let result = if self.is_special(":") {
+            self.advance()?;
+            self.type_tuple()?
+        } else {
+            vec![]
+        };
+        self.expect_newline()?;
+        let mut locals = vec![];
+        let mut instrs = vec![];
```

Comment:
Note that if there are no instructions (empty-bodied function), then it shows up as "native" in the current disassembler output. This may just be a disassembler issue, in which case, it will likely be fixed when we have the new matching disassembler.